### PR TITLE
Add upload mechanism for Linux boards (v2)

### PR DESCRIPTION
**Pull request by lucasdemarchi**
Continuation of https://github.com/ArduPilot/ardupilot/pull/6379

### CODE REVIEW

**Comment by OXINARF** [comment]
I don't quite understand the need for _localinstall_. If I'm getting it correctly it simply copies the built binaries from *build/<board>/bin* to *build/<board>/install*. Why do we need that?

**Comment by lucasdemarchi** [comment]
It actually runs the equivalent of `./waf install` that is how packages are created in Linux. We want to use the same thing for a "temporary local install"... Right now you are right that it just copies some binaries, but it does so respecting --prefix for example.

**Comment by OXINARF** [comment]
> It actually runs the equivalent of ./waf install that is how packages are created in Linux.

What does *localinstall* do differently than the normal *install*?

> it does so respecting --prefix for example

Well, I think it would be a bit weird to install into *\<prefix\>/\<board\>/install*, but OK.

**Comment by lucasdemarchi** [comment]
> What does localinstall do differently than the normal install?

install to a temporary directory, i.e. the equivalent of `DESTDIR=build/<board>/install/ ./waf install`. Then this dir serves as the root dir for rsync, with the proper dir hierarchy. Try this:

```
./waf configure --board navio2
./waf
./waf localinstall
tree build/navio2/install
build/navio2/install
└── usr
    └── local
        └── bin
            ├── antennatracker
            ├── arducopter
            ├── arducopter-heli
            ├── arduplane
            ├── ardurover
            └── ardusub

3 directories, 6 files
```

Note that the default prefix is /usr/local, intended for us to upload the binaries for test.

**Comment by OXINARF** [comment]
> Note that the default prefix is /usr/local, intended for us to upload the binaries for test.

Ah that's what I was missing. Honestly that's even worse: on Windows the default is a temp dir, not `/usr/local` and the user should set the rsync destination, not the build system, there is not even a guarantee the target system has a usable `/usr/local/bin` folder.

**Comment by lucasdemarchi** [comment]
As I said this is intended for ease **developer** testing. People can set the prefix if they want to match whatever configuration they have on the other end.

**Comment by lucasdemarchi** [comment]
I went ahead and merged it. This is not really focusing on Windows and I wouldn't recommend compiling for Linux boards on windows now. We can improve it on top.

**Comment by hiro2233** [comment]
prefix is not a temporary dir, it's developer-user predefined path dir, and i like. It will be great if waf have build, host and target options too.

File: BUILD.md [hunk 1]
```diff
@@ -68,15 +68,25 @@ list some basic and more used commands as example.
 * **Upload or install**
 
     Build commands have a `--upload` option in order to upload the binary built
-    to a connected board. This option is supported by Pixhawk. The command below
-    uses the `--targets` option that is explained in the next item.
+    to a connected board. This option is supported by Pixhawk and Linux-based boards
```
Comment (guludo) [inline BUILD.md]: Missing a period here.

File: BUILD.md [hunk 2]
```diff
@@ -68,15 +68,25 @@ list some basic and more used commands as example.
...
-    install command, which will install to a certain directory. This can be
+    On Linux you need to first configure the IP of the board you are going to upload
+    to. This is done on configure phase with:
+
+    ```sh
+    ./waf configure --board <board> --rsync-dest <destination>
+    ```
+
+    What this does is to install to a temporary location and calling
+    `sync <temp_install_location>/ <destination>`
```
Comment (guludo) [inline BUILD.md]: rsync

**Comment by OXINARF** [comment]
@lucasdemarchi 

> As I said this is intended for ease developer testing.

And you do that by having developers needing to use a prefix option when they already have to set the rsync destination? That makes no sense.

> I went ahead and merged it.

I was quite surprised by that. Next time please write that you'll merge it no matter what so that I don't have to lose my time over it, I definitely have other things to attend to.

> This is not really focusing on Windows and I wouldn't recommend compiling for Linux boards on windows now.

You completely missed the point. You are relying on a default value instead of asking the user to say where he wants it.
Also, there is nothing wrong with compiling on Windows, works perfectly fine.

**Comment by guludo** [comment]
> and the user should set the rsync destination, not the build system, there is not even a guarantee the target system has a usable /usr/local/bin folder.
> You are relying on a default value instead of asking the user to say where he wants it.

The user can set rsync destination with `--rsync-dest`, which can be passed during configuration and also during build. In the case of the latter, the value from the configuration is overridden. The default prefix `/usr/local` is from Waf, they did it because it's somewhat common for 3rd-party software in Linux, others might use `/opt` though. In any case, the user can set the prefix she wants, so I don't see much of a problem here. Lastly, I believe Linux boards that, for some reason, need/want to use a different prefix should set *their* default value for the prefix. We could try making that easy by adding a new attribute like `default_install_prefix` for the board class in `boards.py`.

File: BUILD.md (hunk 1 above)
Comment (lucasdemarchi) [inline BUILD.md]: Thanks, I'll add.

File: BUILD.md (hunk 2 above)
Comment (lucasdemarchi) [inline BUILD.md]: Thanks

**Comment by lucasdemarchi** [comment]
> And you do that by having developers needing to use a prefix option when they already have to set the rsync destination? That makes no sense

That's a misunderstand of what prefix actually is. rsync destination is the root directory of the target system or of the filesystem tree where you want to put the binary. prefix is where the binary will be in the tree and @guludo already clarified. This is really the normal naming convention for Linux distributions creating packages that we are extrapolating here.

> I was quite surprised by that. Next time please write that you'll merge it no matter what so that I don't have to lose my time over it, I definitely have other things to attend to.

Sorry you feel that way. I appreciate reviews. On this one I was actually reviewing a nice thing @guludo did and that I was needing for a long time to ease my testings on Linux-based boards. If it's incomplete or it doesn't work for you, as I said it can be improved on top. I tested on a couple of boards and it does for me.

> You completely missed the point. You are relying on a default value instead of asking the user to say where he wants it.

No, I'm not. `prefix` has the default value defined by waf (which is common to most other build  systems) and can be set by the user.

> Also, there is nothing wrong with compiling on Windows, works perfectly fine.

It's the major reason why waf didn't replace make yet. But it's also offtopic so I suggest to discuss it on an issue.

**Pull request merged**